crates/task-impls/src/quorum_vote/drb_computations.rs

```diff
@@ -0,0 +1,105 @@
+use std::collections::BTreeMap;
+
+use hotshot_types::{
+    drb::{compute_drb_result, DrbResult, DrbSeedInput},
+    traits::node_implementation::NodeType,
+};
+use tokio::{spawn, task::JoinHandle};
+
+/// Helper struct to track state of DRB computations
+pub struct DrbComputations<TYPES: NodeType> {
+    /// Stored results from computations
+    results: BTreeMap<TYPES::Epoch, DrbResult>,
+
+    /// Currently live computation
+    task: Option<(TYPES::Epoch, JoinHandle<DrbResult>)>,
+
+    /// Stored inputs to computations
+    seeds: BTreeMap<TYPES::Epoch, DrbSeedInput>,
+}
+
+impl<TYPES: NodeType> DrbComputations<TYPES> {
+    #[must_use]
+    /// Create a new DrbComputations
+    pub fn new() -> Self {
+        Self {
+            results: BTreeMap::new(),
+            task: None,
+            seeds: BTreeMap::new(),
+        }
+    }
+
+    /// If a task is currently live AND has finished, join it and save the result.
+    /// Does not block if a live task has not finished.
+    pub async fn try_join_task(&mut self) {
+        if let Some((epoch, join_handle)) = &mut self.task {
+            if join_handle.is_finished() {
+                match join_handle.await {
+                    Ok(result) => {
+                        self.results.insert(*epoch, result);
+                    }
+                    Err(e) => {
+                        tracing::error!("error joining DRB computation task: {e:?}");
+                    }
+                }
+                self.task = None;
+            }
+        }
+    }
+
+    /// If a task is currently live AND has finished, join it and save the result.
+    /// If a task is currently live and NOT finished, abort it.
+    /// self.task will always be None after this call.
+    pub async fn join_or_abort_task(&mut self) {
+        if let Some((epoch, join_handle)) = &mut self.task {
+            if join_handle.is_finished() {
+                match join_handle.await {
+                    Ok(result) => {
+                        self.results.insert(*epoch, result);
+                    }
+                    Err(e) => {
+                        tracing::error!("error joining DRB computation task: {e:?}");
+                    }
+                }
+            } else {
+                join_handle.abort();
+            }
+            self.task = None;
+        }
+    }
+
+    /// Starts a new task. If a previous task is currently live, will attempt to join it (if finished), or abort it.
+    pub async fn start_new_task(&mut self, epoch: TYPES::Epoch, drb_seed_input: DrbSeedInput) {
```

Comment:
We can get `drb_seed_input` from `self.seeds` so there's no need to have this as an input.